MSC1929: Homeserver Admin Contact and Support page (#1929)
matrix-org/matrix-spec-proposals: proposals/1929-admin-contact.md

# MSC 1929 Homeserver Admin Contact and Support page

Currently, contacting a homeserver admin is difficult because you need to have insider knowledge
of who the admin actually is. This proposal aims to fix that by specifying a way to add contact details
of admins, as well as a link to a support page for users who are having issues with the service.

This proposal aims to fix https://github.com/matrix-org/matrix-doc/issues/484

## Proposal

> The proposal suggests adding a new endpoint: `.well-known/matrix/support`.

Reviewer: It should be made clear whether it is served on the main matrix domain or the delegated matrix server(s) or both.

Reviewer 2: I would say on the same well-known as the `.../server` endpoint.

Reviewer 3: For this MSC to be useful, it'd have to be off the server name, like @grinapo mentions.

Reviewer 4: ```suggestion
The proposal suggests adding a new endpoint: `https://{hostname}/.well-known/matrix/support`,
where `hostname` is the server name, without the port number.  This is the same as what is used
when performing server discovery.
```

Reviewer 5: Sorry for nitpick: Shouldnt it be `server_name` rather than `hostname`? :) Since server_name I think is well defined in spec already.

Reviewer 3: Technically it is the hostname, similar to https://spec.matrix.org/v1.8/client-server-api/#well-known-uri

The response format should be:

```javascript
{

>     admins: [{

Reviewer 5: I am kind of wondering: Would it make sense to have a field "allows automatic reports"? As I am wanting to write a web interface for mjolnir and other moderation and having it also look for this would be helpful. However I just cant assume if an admin wants to be contacted or if an automated "User xy was banned in abc for spam" would be sufficient for an admin. Having a flag for this would be helpful so the admin can decide on this.

On the other hand this may actually even be a MSC building on top so it also can specify metadata in the event for further automation. Not quite sure currently

Reviewer 2: As for myself I would probably create a **room** and join there, being able to fine-tune its notification parameters or automagic handling; and I would provide it there.

Reviewer 3: Automated report handling is likely best discussed in another MSC. There's some plans being formed to figure out such systems, and it'd be best to not have this MSC get stuck in those conversations.

>         matrix_id: "@admin:domain.tld",

Reviewer 6: I'm bikeshedding but if it's `email` (rather than `email_address`) it should be `matrix`, no?

Reviewer 3: *resolving because I think this was handled in other ways*

>         email_address: "admin@domain.tld",

Reviewer 5: I would expect that to be done by adding multiple admins in that case, or just use mailinglists/mail forwarding. I don't think this is something that should be solved within this.

>         email_address: "admin@domain.tld",

Reviewer 5: I think only the msc author can mark it as complete :)

        role: "admin" # If omitted, the default will be "admin"

>     },

Reviewer 2: May it be useful to specify a room-id (which could be joined by multiple/all/varying admins)? They may (and probably should) use foreign servers to watch the room.

Author: RoomId (possibly alias is better) is a good shout, though I don't know if we plan to have a general "landing" room in other places and this may be a duplication of effort. Will leave that question open to other reviewers.

Reviewer 7: Although I think RoomID might be a possible solution I don't think this is sth. people will use in general as I think this contact will mainly be used to notify about updates, spam or security issues.
These are information I as an admin would not want to be shared with everybody.

As an alternative I think a group might fit here as well. This could then contain multiple users which take the admin role. This especially can better administered inside of matrix

Author: I'd love to specify a group, but I'd keep that in mind for a future extension when we have groups in the spec. And certainly the reason for the roles given in the proposal is to specify the sensitivity for an account, amongst other things.

Reviewer 3: I'd suggest that this get pushed to a future MSC.

Reviewer 2: Or use a way which makes it possible to flexibly use any any of the identifiers, may it be email, mxid, room_id, or the future roaming_or-whatever_id replacing mxid.

```json
 { "contact": [  
    { "role": "admin", "type": "email", "id": "admin@example.org" }, 
    { "role": "admin", "type": "mxid", "id": "@admin@example.eu" }, 
    { "role": "admin", "type": "room_id", "id": "!98765432ytuihuiop" },
    { "role": "network", "type": "pk_id", "id": "8AF70C62FB01667548E7FBA4A7143B706592813D:origin.mars" },
   ] }
```

rooms can be spaces as well.

Reviewer 3: I'm not really convinced that we need *more* identifiers, honestly. Emails and user IDs cover the vast majority of scenarios, and for reporting all other concerns we may very well need a different system entirely. A future MSC can take responsibility for considering additional details and reporting flows.

In general, fewer options is better to provide a consistent expectation for Matrix users.

Reviewer 2: I tried to be forward compatible, but it's not that important.

Reviewer 8: > I'd love to specify a group, but I'd keep that in mind for a future extension when we have groups in the spec. And certainly the reason for the roles given in the proposal is to specify the sensitivity for an account, amongst other things.

We have rooms that allow high powerlevels to send messages, is there a way to make a room that unprivileged users can send to, but only high powerlevels will be able to receive those messages? I know that Matrix != IRC, but that was a thing you could do on some IRC servers for abuse reporting rooms.

Reviewer 3: Send-only rooms would definitely be a future MSC (not currently possible)

Author: I'm also of the belief that this is future MSC stuff for the moment. I think you could achieve this with a contact bot that puts you into a abuse room perhaps, but for now I think let's run with what we have.

>     },

Reviewer 2: In the course of [MSC2063](https://github.com/matrix-org/matrix-doc/pull/2063) (server info for automated lists) I have realised (w/ help of @Half-Shot ) that static server information would be much better to provide here, namely:
* server_name: human-readable server name
* server_description: a human-readable longer description, for example about speciality local rooms, local rules, restrictions
* server_country: the home country or the target country of the server, if any

This is all static, manually defined data about the server, which would be better situated in the static `.well-known` than dynamically generated by a matrix endpoint.

Reviewer 3: As a support endpoint, I'm not sure this information is relevant here. Other MSCs, like 2063, are better suited for general server information, I think.

    {
        email_address: "security@domain.tld",
        role: "security"
    }],

>     support_page: "https://domain.tld/support.html"

Reviewer 9: Adding a `status_page` field that could be included in client UI when connection to the homeserver is lost could be helpful as well.

Reviewer 3: see https://github.com/matrix-org/matrix-spec-proposals/pull/1929/files#r1365832870

> }

Reviewer 3: see https://github.com/matrix-org/matrix-spec-proposals/pull/1929/files#r1365832870

> }

Reviewer 3: see https://github.com/matrix-org/matrix-spec-proposals/pull/1929/files#r1365832870

> }

Reviewer 4: > see https://github.com/matrix-org/matrix-spec-proposals/pull/1929/files#r1365832870

This link doesn't seem to be working any more.

> }

Reviewer 3: Thanks GitHub. It's essentially saying the idea is out of scope - other MSCs are better positioned to handle the functionality.

> }

Author: Agree. I think this is future MSC territory.

> }

Author: Sorry, again I think this is good for future MSCs. This sounds like a good idea though.

> ```javascript
> {
>     admins: [{
>         matrix_id: "@admin:domain.tld",
>         email_address: "admin@domain.tld",
>         role: "admin" # If omitted, the default will be "admin"
>     },
>     {
>         email_address: "security@domain.tld",
>         role: "security"
>     }],
>     support_page: "https://domain.tld/support.html"
> }
> ```

Reviewer 10: I assume this is supposed to be JSON?

```suggestion
```json
{
    "admins": [{
        "matrix_id": "@admin:domain.tld",
        "email_address": "admin@domain.tld",
        "role": "admin" # If omitted, the default will be "admin"
    },
    {
        "email_address": "security@domain.tld",
        "role": "security"
    }],
    "support_page": "https://domain.tld/support.html"
}
```
```

The `mxid` and `email` do NOT need to have the same domain as the homeserver. It is expected that
a server will have a "backup" contact address if the server is down, like an email.

> Entries may have a `mxid` OR an `email`, but at least one MUST be specified.

Reviewer 4: Would `phone` be useful too?  Also, I think having an (optional) name could be helpful.  (Also, I'm not too fond of calling it `mxid`, and IMHO would be better calling it `matrix_id`.)

Author: +1 to matrix_id, although I don't know what's canonical form for the spec currently.

We could have phone, but I really can't see anyone phoning/texting a number for support?

`role` is an informal description of what the address(es) are used for. The only two specified in this
proposal are "admin" and "security." Admins are a catchall user for any queries, where security is intended

> for sensitive requests.

Reviewer 3: [non-blocking concern] it may be valuable to namespace the field value here

Author: Works for me.

> `admins` is optional, but recommended.

Reviewer 11: What was the reasoning behind the choice of `admins`?

Having an admin of type (role) `admin` and an admin of type `security` seems a bit weird and it doesn't seem flexible in terms of further roles (maybe a helpdesk contanct?).

My first instinct would have been to name them `contacts` and have the role specify their function.

Author: Going with `contacts`

`support_page` is a optional property to specify a affiliated page of the homserver to give users help

> specific to their installation.

Reviewer 4: Would this be used, for example, by clients to automatically show a link to users to help them find support?  Are there other use cases that you're thinking of?

If it's mainly for clients to present the information to users, maybe it would be better to put this is `.well-known/client`, so that clients only need to fetch one `.well-known` file, rather than two.

Author: Yeah I had imagined this would be presented to the client in case of failure, or for troubleshooting steps with logins. I'm not adverse to taking this out of `./support` for now and making support pages be part of a separate proposal.

Author: >If it's mainly for clients to present the information to users

This is more server facing for admins to complain to other admins, but doubles as a local support mechanism too.

## Alternative solutions

Hardcode a given user localpart that should be used as an admin address.
 - The account would need to either internally redirect messages intended for @admin:domain.tld to another account(s)

>  - OR require an admin to regularly sign into this special account to check for messages. Neither of which is useful.

Reviewer 6: `Neither is useful` is a pretty strong statement that certainly needs clarification for the first part. I understand that redirection does have its shortcomings but the proposed primary solution has them too. Consider the case of two timeshifting admins, e.g. - does it mean that `.well-known` will have to be switched twice a day, and clients have to somehow discover that? And what if the support line involves load balancing across 10 people?

Quite frankly, I don't see how far the proposed solution is from hardcoding a certain user localpart. Once the user MXID is specified in `.well-known`, you effectively declare that (at this arbitrary period of time) this and only this Matrix user is supposed to receive all support requests, right? Now how many people would want to re-use their primary account for admin tasks? And if it's not primary, doesn't it force them into option 2 of regularly signing into the admin account?

Author: >Consider the case of two timeshifting admins, e.g. - does it mean that .well-known will have to be switched twice a day, and clients have to somehow discover that?

No, you would define both admins in the list. If we think it's useful, you could optionally define a timezone.

This data is meant to be static, and rarely changed when staff shuffle around which I would expect to be uncommon. In the case of irregular support systems, there is the option of specifying a support page. I don't intend to spec a complete format for every type of support system people might use, but the most common one is just specifying a contact address of one or more admins.

>And what if the support line involves load balancing across 10 people?

*Then* I would suggest listing one mxid in the entry list, and writing some sort of bot or ticketing system to handle it. That's certainly a uncommon option for the vast amount of homeservers out there, and not part of this proposal.

>Quite frankly, I don't see how far the proposed solution is from hardcoding a certain user localpart

Fundamentally, you could specify `@_admin:domain.tld` in the entry list if you wanted and it would function the same way as a hardcode. The proposal gives the option to be smarter than that and list the actual admin, or an email address. 

Hardcoding a support mxid means more work on the homeserver to direct messages to the right place, or for admins to now sign into two accounts everywhere to catch messages.

>Now how many people would want to re-use their primary account for admin tasks?

The vast vast vast majority of homeserver admins out there run small homeservers and probably won't be getting a lot of admin queries. For the few that do have a large amount of support traffic, specify a second account.

In general I don't see a reason why you cannot specify a config like:

```javascript
{
    admins: [{ matrix_id: "@admin:domain.tld" }],
}
```

if you wanted to have a dedicated admin account, but the proposal doesn't force people into setting one up.

Reviewer 2: Synapse (or general homeserver) at installation/configuration stage may ask for the admin id, or may set it to user#1 unless the admin's changed it? So a hardcoded address can redirect to user#1 by default, which would be changeable.

Author: Yeah, it could do. I'd probably say that the operator needs to consciously elect a admin user, rather than it defaulting to a user to avoid us accidentally leaking lots of private userids.

Author: @KitsuneRal thoughts?

Reviewer 6: It only took me 3 years to get back to this thread - anyway, now that I'm older and wiser I think I can agree with Half-Shot's ideas.

Specify the same content inside a homeserver endpoint, rather than use `.well-known`.
 - This requires the homeserver to be up or responsive, which might be not very useful if trying to report issues with
   connectivity.
   
Use vCards.
 - vCards would add bloat, as the vast majority of a vcards contents is not useful for contacting an admin.

## Security considerations

> 

Reviewer 2: Isn't it a security consideration that _admin email is available publicly processable by automatic ("spambotty") means_?
I don't know about matrix-targetting spambots but matrix-ids may be a valuable asset in the near future, when matrix start to dominate the communication of the world.

Author: So this is a problem regardless of medium, although I admit that more people are going to attack email once they notice the new format become a standard.

Realistically I would expect the homeserver/email provider to do the usual spam checks/ reputation work to ensure that spam isn't making it's way to users.

Fundamentally we need to provide a way to allow users from any server to contact the admin of another server and the two ways of doing it are providing a standard address (e.g. `@admin:domain.tld` ) or allowing it to be custom as in this proposal. This proposal is not much better, but at least requires an ounce more effort of a spammer to go and find your address. My final point is that you could carry this style of attack out simply by joining a large public room, so I wouldn't say this proposal is making things any easier

Tl;Dr - I feel like the spam problem is going to happen whatever mechanism we provide, and I don't have a good answer other than other implementations adding sensible precautions.

Reviewer 3: There are some options with automated reporting structures, though as discussed elsewhere this sort of scope creep is ideally left to a future MSC. For example, the spec could have endpoints to report abuse, where the backend happens to send an email or use a different system. This effectively protects the identifiers from direct spam.

There's obviously still the problem of using the dedicated endpoints to also spam reports, but at least unrelated email spam wouldn't make it to the address.

In the meantime, I think this MSC is still net-positive as a change.

Author: I think this concern is noted, but we're going ahead for the moment. Future MSCs may offer better mediums.

Reviewer 2: The original comment was about expanding the _Security considerations_ section with exatcly what have been said here. ;-)
It does not need to be fixing (that'd be a future MSC), it just needs to be documented. But I'm still not sure how I'd phrase it not to be overly verbose.

> # MSC1929 Homeserver Admin Contact and Support page

Reviewer 5: Draupnir Implementation has been started at https://github.com/the-draupnir-project/Draupnir/pull/150

It adds a `queryAdmin <entity | domain>` command which allows querying this data in a user-friendly way. The implementation is not tested at the time of writing this.

> # MSC1929 Homeserver Admin Contact and Support page

Reviewer 3: @TheArcaneBrony says:

> Additional field request: `contact_all` (boolean) Motivation: special cases where all admins/accounts are recommended to be contacted at once, rather than a single admin, for cases where individual contact may be futile. Examples:
> 
> * Single-user homeservers, where a user has accounts across multiple homeservers and may not check all accounts frequently
> * Cases like rory.gay where all admins are part of one physical body and the a given admin may literally not be around for weeks, months or years (eg. `@rory:rory.gay` and `@root:rory.gay` have not been around for months)

(please use threads)

Author: I'm feeling that this would be better off as a future MSC. I get that this might chafe for some admin teams, but perhaps this is where a forwarding bot makes sense. I worry this might bloat the spec, or at least the v1 of the spec.

> # MSC1929 Homeserver Admin Contact and Support page

Reviewer 12: Is there an alternative for servers which do not run a HTTP(S) server on their main domain and use a `SRV` record instead?

Reviewer 3: Not typically for MSCs like this. We're moving away from SRV records for federation lookups, and this is before the federation lookup anyways.

Reviewer 13: > We're moving away from SRV records for federation lookups

[are we](https://github.com/matrix-org/matrix-spec-proposals/pull/3922#discussion_r1022527789) ?

Reviewer 3: It's up for reconsideration thanks to MIMI :)

Reviewer 2: It doesn't say it's **not** obsolete just that nobody took the time to say it is... :shrug:

Author: I'm torn on this one. We really want this to be widespread since the whole point is that open servers identify their admins for the purpose of good faith moderation and support. However, I feel like it would be a departure. The other problem is that for clients, they would have no reasonable way to do a SRV lookup and so would be stuck.

I think regretfully I'd prefer it if we kept this as .well-known only and hope most of the network can adopt `.well-known`?

Reviewer 3: We shouldn't tie this MSC's acceptance to server discovery, in my opinion. Future MSCs are welcome to discuss a non-HTTPS way of communicating this information, but for the purposes of this MSC the use of SRV records feels irrelevant to the scope.

Author: I think we're in agreement here then for now :+1:. Resolved?

Currently, contacting a homeserver admin is difficult because you need to have insider knowledge
of who the admin actually is. This proposal aims to fix that by specifying a way to add contact details
of admins, as well as a link to a support page for users who are having issues with the service.

This proposal aims to fix https://github.com/matrix-org/matrix-doc/issues/484

## Proposal

The proposal suggests adding a new endpoint: `.well-known/matrix/support`.

The response format should be:

> ```json5
> {
>     {
>         "contacts": [{
>             "matrix_id": "@admin:domain.tld",
>             "email_address": "admin@domain.tld",
>             "role": "admin" // If omitted, the default will be "admin"
>         },
>         {
>             "email_address": "security@domain.tld",
>             "role": "security"
>         }],
>         "support_page": "https://domain.tld/support.html"
>     }
> }

Reviewer 3: This is double nested for some reason:
```suggestion
```json5
{
    "contacts": [{
        "matrix_id": "@admin:domain.tld",
        "email_address": "admin@domain.tld",
        "role": "admin" // If omitted, the default will be "admin"
    },
    {
        "email_address": "security@domain.tld",
        "role": "security"
    }],
    "support_page": "https://domain.tld/support.html"
}
```

Reviewer 2: What do you think about my suggestion generalising contact methods, using `type` and `id` instead or matrix_id, email_id, room_id, whatever id?

```json

{ "contacts": [  
    { "role": "admin", "type": "email", "id": "admin@example.org" }, 
    { "role": "admin", "type": "mxid", "id": "@admin@example.eu" }, 
    { "role": "admin", "type": "room_id", "id": "!98765432ytuihuiop" },
    { "role": "security", "type": "pk_id", "id": "8AF70C62FB01667548E7FBA4A7143B706592813D:origin.mars" },
   ] }
```

or if it's useful to group by person/entity:

```json
{ "contacts": [  
    { "role": "admin", [
       {  "type": "email", "id": "admin@example.org" }, 
       {  "type": "mxid", "id": "@admin@example.org" }
     ]},
    { "role": "security", [
       {  "type": "room", "id": "!87645432oliuyiuyt:woodoo.venus" }, 
       {  "type": "mxid", "id": "@abuse@example.org" }
     ]}
   ]}

Reviewer 3: I've replied in the existing thread as it's off topic here: https://github.com/matrix-org/matrix-spec-proposals/pull/1929#discussion_r1367175717

Reviewer 13: if you're fixing the formatting, can you indent the content of `contacts` please

```

The `matrix_id` and `email_address` do NOT need to have the same domain as the homeserver. It is expected that
an admin will have a "backup" contact address if the server is down, like an email or alternative mxid on a different homeserver.

Entries may have a `matrix_id` OR an `email_address`, but at least one MUST be specified.

`role` is an informal description of what the address(es) are used for. The only two specified in this
proposal are "admin" and "security." Admins are a catchall user for any queries, where security is intended

> for sensitive requests.

Reviewer 6: Not sure why it should be required - the default "admin" value is implied, no?

> for sensitive requests.

Reviewer 3: Implied != specified :D 

(I did not read this as default being "admin")

> for sensitive requests.

Reviewer 14: The response format says

> "role": "admin" // If omitted, the default will be "admin"

If it's required, it shouldn't say it has a default when omitted

> for sensitive requests.

Author: I think we should just make this explicit to be honest. There doesn't seem to be a whole lot of value in implicit defaults.

> for sensitive requests.

Author: I've made this explicit.

> `admins` is optional, but recommended.

Reviewer 5: Just realized: Both support_page and admins/contacts are optional. Shouldn't the MSC specify that one of them is required. Aka: If you omit admins it should have a support_page and if that's missing it should have the admins/contacts data? It would be pretty useless if there is just an empty json.

Reviewer 5: > > It would be pretty useless if there is just an empty json.
> 
> It could signify "no contacts" while still supporting this MSC. Useless, but is it needed to be specified? (For the sake of simplicity of the MSC.)

I think this defeats the purpose of this MSC at the core. :)

Author: https://github.com/matrix-org/matrix-spec-proposals/pull/1929/files#r1394489570 is hopefully some wording for this.

The response format should be:

```json5
{
    "contacts": [
        {
            "matrix_id": "@admin:domain.tld",
            "email_address": "admin@domain.tld",
            "role": "admin" // If omitted, the default will be "admin"
        },
        {
            "email_address": "security@domain.tld",
            "role": "security"
        }
    ],
    "support_page": "https://domain.tld/support.html"
}
```

The `matrix_id` and `email_address` do NOT need to have the same domain as the homeserver. It is expected that
an admin will have a "backup" contact address if the server is down, like an email or alternative mxid on a different homeserver.

Entries may have a `matrix_id` OR an `email_address`, but at least one MUST be specified.

`role` is an informal description of what the address(es) are used for. The only two specified in this
proposal are "admin" and "security." Admins are a catchall user for any queries, where security is intended
for sensitive requests.

`contacts` is optional, but recommended.

`support_page` is an optional property to specify a affiliated page of the homserver to give users help
specific to the homeserver, like extra login/registration steps.

> 

Author: ```suggestion

At least one valid key should be provided. This means `contacts` should have at least one entry, or the `support_page` should be defined. An empty object is not considered valid, however both `contacts `and `support_page` may be specified together. 

```

when performing server discovery.

The response format should be:

```json5
{
    "contacts": [
        {
            "matrix_id": "@admin:domain.tld",
            "email_address": "admin@domain.tld",
            "role": "m.role.admin"
        },
        {
            "email_address": "security@domain.tld",
            "role": "m.role.security"
        }
    ],
    "support_page": "https://domain.tld/support.html"
}
```

The `contacts` array is optional, but recommended.

The `matrix_id` and `email_address` do NOT need to have the same domain as the homeserver. It is expected that
an admin will have a "backup" contact address if the server is down, like an email or alternative mxid on a different homeserver.

Entries may have a `matrix_id` OR an `email_address`, but at least one MUST be specified.

`role` is an informal description of what the address(es) are used for. The only two specified in this
proposal are `m.role.admin` and `m.role.security`. 

 - `m.role.admin` is a catch-all user for any queries.
 - `m.role.security` is intended for sensitive requests

Implementors may use custom values for role for other purposes, but it's suggested

> that the value be namespaced to prevent collisions. A value for `role` MUST be specified.

Reviewer 13: Could you specify a grammar for the value of  `role` please. I recommend https://spec.matrix.org/v1.8/appendices/#common-namespaced-identifier-grammar.

Reviewer 3: Applied manually in https://github.com/matrix-org/matrix-spec-proposals/commit/676939659c103627a4de8fa7fe9c39693f6d1340

> `support_page` is an optional property to specify a affiliated page of the homserver to give users help

Reviewer 13: ```suggestion
`support_page` is an optional property to specify an affiliated page of the homeserver to give users help
```

Reviewer 3: Applied manually in https://github.com/matrix-org/matrix-spec-proposals/commit/676939659c103627a4de8fa7fe9c39693f6d1340

        {
            "matrix_id": "@admin:domain.tld",
            "email_address": "admin@domain.tld",

>             "role": "org.matrix.admin"

Reviewer 3: if namespaced, we'd use `m.`

Author: duh, thanks.
